readme.md
# Bingo Kata

This project is the implementation of a US Bingo, taken from this [test](http://www.tddbuddy.com/katas/Bingo%20Kata.pdf).

## Prerequisites

This project is using:
  - [Node.js](https://nodejs.org/es/), v 14.16.0 
  - [Yarn](https://yarnpkg.com/), v 1.22.10

Clone this project and run the following commands in your terminal:

```console
$ yarn
```

## Usage

There are three sections to test separately, by running the followings commands:

- To test Calling Bingo Numbers section:
  ```
  $ yarn callingBingoNumbers
  ```

- To test Generating Bingo Cards section:
  ```
  $ yarn generatingBingoCards
  ```

- To test Checking Bingo Cards section:
  ```
  $ yarn checkingBingoCards
  ```

## Author
-   **Anthony Luzquiños** - _Initial Work_ - _Documentation_ - [AnthonyLzq](https://github.com/AnthonyLzq).

changelog.md
# Bingo Kata

## Version 0.1.0

- Implemented:
  - Project initialization
  - Calling Bingo Numbers section

## Version 0.2.0
- Implemented:
  - Generating Bingo Cards
  - Updated documentation

## Version 0.3.0
- Implemented:
  - Checking Bingo Cards
  - Fixed changelog
  - Updated documentation
  - Prettifying files


src/checkingBingoCards.test.js
// TESTING CHECKING BINGO CARDS
const feature3 = require('./features/checkingBingoCards')
const { validatingPlay } = feature3

test('Winner', () => {
  const result = validatingPlay()

  expect(result).toEqual('winner')
})

test('Loser', () => {
  const result = validatingPlay('loser')

  expect(result).toEqual('loser')
})


src/callingBingoNumbers.test.js
// TESTING CALLING BINGO NUMBERS

const feature1 = require('./features/callingBingoNumbers')
const {
  callANumber,
  createBingoNumbers,
  getCurrentBingoArray,
  resetBingoNumbers
} = feature1

// Testing if every number is being called once
test('Every number is called once', () => {
  const aux = createBingoNumbers()
  const result = []
  for (let i = 0; i < 75; i++) result.push(callANumber().next().value)

  expect(aux.sort()).toEqual(result.sort())
})

// Testing no more numbers available
test('All the numbers were called', () => {
  expect(() => callANumber().next().value).toThrow(
    'All the numbers were called'
  )
})

// Testing we can re populate the array of numbers
test('We have numbers from 1 to 75 inclusive again', () => {
  resetBingoNumbers()
  const aux = createBingoNumbers()
  const result = getCurrentBingoArray()

  expect(aux.sort()).toEqual(result.sort())
})


src/generatingBingoCards.test.js
// TESTING GENERATING BINGO CARDS

const feature2 = require('./features/generatingBingoCards')
const { bingoCardBuilder } = feature2

test('Every card has 24 numbers', () => {
  const bingoCard = bingoCardBuilder.getBingoCard()
  const numberOfCards =
    bingoCard.B.length +
    bingoCard.I.length +
    bingoCard.N.length +
    bingoCard.G.length +
    bingoCard.O.length

  expect(numberOfCards).toEqual(24)
})

test('Every card has numbers between its bounds', () => {
  const bingoCard = bingoCardBuilder.getBingoCard()
  let areTheValuesBetweenBounds = true

  // For the first column
  bingoCard.B.forEach(
    value => areTheValuesBetweenBounds && value >= 1 && value <= 15
  )

  // For the second column
  bingoCard.I.forEach(
    value => areTheValuesBetweenBounds && value >= 16 && value <= 30
  )

  // For the second column
  bingoCard.N.forEach(
    value => areTheValuesBetweenBounds && value >= 31 && value <= 45
  )

  // For the second column
  bingoCard.G.forEach(
    value => areTheValuesBetweenBounds && value >= 46 && value <= 60
  )

  // For the second column
  bingoCard.O.forEach(
    value => areTheValuesBetweenBounds && value >= 61 && value <= 75
  )

  expect(areTheValuesBetweenBounds).toEqual(true)
})


src/features/callingBingoNumbers.js
// CALLING BINGO NUMBERS

// We create a function to generate an array of numbers from 1 to 75 inclusive
const createBingoNumbers = () => {
  const bingoNumbersArray = []
  for (let i = 1; i <= 75; i++) bingoNumbersArray.push(i)

  return bingoNumbersArray
}

// We create an array with all the numbers from 1 to 75 inclusive
let BINGO_NUMBERS_ARRAY = createBingoNumbers()

// Generate a caller
function* callANumber() {
  if (BINGO_NUMBERS_ARRAY.length > 0) {
    BINGO_NUMBERS_ARRAY.sort(() => Math.random() - 0.5)

    yield BINGO_NUMBERS_ARRAY.splice(0, 1)[0]
  } else throw new Error('All the numbers were called')
}

// Create a function to reset the array in case we need it
const resetBingoNumbers = () => (BINGO_NUMBERS_ARRAY = createBingoNumbers())

const getCurrentBingoArray = () => BINGO_NUMBERS_ARRAY

module.exports = {
  callANumber,
  createBingoNumbers,
  getCurrentBingoArray,
  resetBingoNumbers
}


src/features/generatingBingoCards.js
// Generating Bingo Cards

class BingoCard {
  constructor(B, I, N, G, O) {
    this.B = B
    this.I = I
    this.N = N
    this.G = G
    this.O = O
  }
}

class BingoCardGenerator {
  getBingoCard() {
    const B = this._fillArray(1, 15)
    const I = this._fillArray(16, 30)
    const N = this._fillArray(31, 45, 4)
    const G = this._fillArray(46, 60)
    const O = this._fillArray(61, 75)

    return new BingoCard(B, I, N, G, O)
  }

  _randomNumberGeneratorBetweenBounds(lowerBound, upperBound) {
    return Math.floor(Math.random() * (upperBound - lowerBound) + lowerBound)
  }

  _fillArray(lowerBound, upperBound, limit = 5) {
    const aux = []
    for (let i = 0; i < limit; i++)
      aux.push(this._randomNumberGeneratorBetweenBounds(lowerBound, upperBound))

    return aux
  }
}

module.exports = {
  bingoCardBuilder: new BingoCardGenerator()
}


src/features/checkingBingoCards.js
// Checking bingo cards

const { callANumber } = require('./callingBingoNumbers')
const { bingoCardBuilder } = require('./generatingBingoCards')

const validatingPlay = (result = 'winner') => {
  const bingoCard = bingoCardBuilder.getBingoCard()
  let numbersLeft =
    bingoCard.B.length +
    bingoCard.I.length +
    bingoCard.N.length +
    bingoCard.G.length +
    bingoCard.O.length

  while (true) {
    const number = callANumber().next().value

    if (number >= 1 && number <= 15 && bingoCard.B.includes(number)) {
      bingoCard.B.splice(bingoCard.B.indexOf(number), 1)
      numbersLeft--
    } else if (number >= 16 && number <= 30 && bingoCard.I.includes(number)) {
      bingoCard.I.splice(bingoCard.I.indexOf(number), 1)
      numbersLeft--
    } else if (number >= 31 && number <= 45 && bingoCard.N.includes(number)) {
      bingoCard.N.splice(bingoCard.N.indexOf(number), 1)
      numbersLeft--
    } else if (number >= 46 && number <= 60 && bingoCard.G.includes(number)) {
      bingoCard.G.splice(bingoCard.G.indexOf(number), 1)
      numbersLeft--
    } else {
      bingoCard.O.splice(bingoCard.O.indexOf(number), 1)
      numbersLeft--
    }

    if (result === 'winner' && numbersLeft === 0) return 'winner'
    else if (result === 'loser' && numbersLeft === 1) return 'loser'
  }
}

module.exports = {
  validatingPlay
}
